core/src/main/scala/pureconfig/ConfigConvert.scala

```diff
@@ -155,27 +157,44 @@ object ConfigConvert extends LowPriorityConfigConvertImplicits {
   case class NoValidCoproductChoiceFound(config: ConfigValue)
     extends RuntimeException(s"No valid coproduct type choice found for configuration $config.")
 
-  implicit def cNilConfigConvert: ConfigConvert[CNil] = new ConfigConvert[CNil] {
+  implicit def cNilConfigConvert[Wrapped]: WrappedConfigConvert[Wrapped, CNil] = new WrappedConfigConvert[Wrapped, CNil] {
     override def from(config: ConfigValue): Try[CNil] =
       Failure(NoValidCoproductChoiceFound(config))
 
     override def to(t: CNil): ConfigValue = ConfigFactory.parseMap(Map().asJava).root()
   }
 
-  implicit def coproductConfigConvert[V, T <: Coproduct](
-    implicit vFieldConvert: Lazy[ConfigConvert[V]],
-    tConfigConvert: Lazy[ConfigConvert[T]]): ConfigConvert[V :+: T] =
-    new ConfigConvert[V :+: T] {
+  implicit def coproductConfigConvert[Wrapped, Name <: Symbol, V, T <: Coproduct](
+    implicit coproductHint: CoproductHint[Wrapped],
+    vName: Witness.Aux[Name],
+    vFieldConvert: Lazy[ConfigConvert[V]],
+    tConfigConvert: Lazy[WrappedConfigConvert[Wrapped, T]]): WrappedConfigConvert[Wrapped, FieldType[Name, V] :+: T] =
+    new WrappedConfigConvert[Wrapped, FieldType[Name, V]:+: T] {
+
+      override def from(config: ConfigValue): Try[FieldType[Name, V] :+: T] = config match {
+        case null => Failure(CannotConvertNullException)
+
+        case _ => coproductHint.from(config, vName.value.name) match {
+          case Success(Some(hintConfig)) =>
+            vFieldConvert.value.from(hintConfig) match {
+              case Failure(_) if coproductHint.tryNextOnFail(vName.value.name) =>
+                tConfigConvert.value.from(config).map(s => Inr(s))
 
-      override def from(config: ConfigValue): Try[V :+: T] = {
-        vFieldConvert.value.from(config)
-          .map(s => Inl[V, T](s))
-          .orElse(tConfigConvert.value.from(config).map(s => Inr[V, T](s)))
+              case vTry => vTry.map(v => Inl(field[Name](v)))
+            }
+
+          case Success(None) => tConfigConvert.value.from(config).map(s => Inr(s))
+          case Failure(ex) => Failure(ex)
+        }
       }
 
-      override def to(t: V :+: T): ConfigValue = t match {
-        case Inl(l) => vFieldConvert.value.to(l)
-        case Inr(r) => tConfigConvert.value.to(r)
+      override def to(t: FieldType[Name, V] :+: T): ConfigValue = t match {
+        case Inl(l) =>
+          // Writing a coproduct to a config can fail. Is it worth it to make `to` return a `Try`?
```

Comment:
I wonder this too. I always had the idea that `to` has to be total but I see your point here. Using `get` makes sense if you want to throw exception but personally I prefer to wrap `to` into `Try` and just let the user know that `to` can result in a `Failure`. Method signatures are good communication mechanism in particular when they point to possible failures.